reasoner/resolution/resolver/BoundConcludableResolver.java

```diff
@@ -173,128 +122,85 @@ public boolean singleAnswerRequired() {
             return nextAnswer().map(Partial::asCompound);
         }
 
-    }
-
-    protected class BlockedRequestState<ANSWER> extends BoundConcludableRequestState<ANSWER> {
-
-        public BlockedRequestState(Request.Template fromUpstream, AnswerCache<ANSWER> answerCache, boolean deduplicate,
-                                   UpstreamBehaviour<ANSWER> upstreamBehaviour, boolean singleAnswerRequired) {
-            super(fromUpstream, answerCache, deduplicate, upstreamBehaviour, singleAnswerRequired);
+        protected AnswerCache<ANSWER> cache() {
+            return answerCache;
         }
 
-        @Override
-        void sendNextMessage(Trace trace) {
-            Optional<Partial.Compound<?, ?>> upstreamAnswer = upstreamAnswer();
-            if (upstreamAnswer.isPresent()) {
-                if (singleAnswerRequired()) cache().setComplete();
-                answerToUpstream(upstreamAnswer.get(), fromUpstream().createVisit(trace));
-            } else if (cache().isComplete()) {
-                failToUpstream(fromUpstream().createVisit(trace));
-            } else {
-                blockToUpstream(fromUpstream().createVisit(trace), cache().size());
-            }
-        }
+        public abstract void resetCacheSource();
+    }
 
-        @Override
-        void receiveRevisit(Trace trace, Set<Cycle> cycles) {
-            sendNextMessage(trace);
+    class MatchResolutionState extends BoundConcludableResolutionState<ConceptMap> {
+
+        protected MatchResolutionState(Partial<?> fromUpstream, AnswerCache<ConceptMap> answerCache,
+                                       boolean singleAnswerRequired) {
+            super(fromUpstream, answerCache, true, singleAnswerRequired);
         }
 
         @Override
-        void receiveAnswer(Response.Answer fromDownstream) {
-            throw TypeDBException.of(ILLEGAL_STATE);
+        public ConceptMap answerFromPartial(Partial<?> partial) {
+            return partial.conceptMap();
         }
 
         @Override
-        void receiveFail(Response.Fail fromDownstream) {
-            throw TypeDBException.of(ILLEGAL_STATE);
+        public FunctionalIterator<? extends Partial<?>> toUpstream(ConceptMap conceptMap) {
+            return Iterators.single(fromUpstream.asConcludable().asMatch().toUpstreamLookup(
+                    conceptMap, context.concludable().isInferredAnswer(conceptMap)));
         }
 
         @Override
-        void receiveBlocked(Response.Blocked fromDownstream) {
-            throw TypeDBException.of(ILLEGAL_STATE);
+        public void resetCacheSource() {
+            answerCache.setSource(() -> traversalIterator(context.concludable().pattern(), bounds));
         }
     }
 
-    protected class ExploringRequestState<ANSWER> extends BoundConcludableRequestState<ANSWER> implements RequestState.Exploration {
+    static class ExplainResolutionState extends BoundConcludableResolutionState<Partial.Concludable<?>> {
```

Comment:
I agree it's weird, but there are two methods that need fields from the outer scope. In fact we can't do this easily at all because `resetCacheSource` needs to have a different behaviour for the two types of resolutionState.